File: social_model.py
from mesa import Agent, Model
from mesa.time import RandomActivation
from mesa.space import SingleGrid
from mesa.datacollection import DataCollector


class Municipality(Agent):
    def __init__(self, unique_id, has_legislated, model):
        super().__init__(unique_id, model)
        self.has_legislated = has_legislated

    def step(self):

        if self.model.pike_abundance <= self.model.pike_threshold:
            if self.model.years_after_threshold_breach == 0 and not self.has_legislated:
                print("start legislation process")
                self.model.years_after_threshold_breach += 1
                self.has_legislated = True

        if self.model.years_after_threshold_breach == 1:
            self.inform_neighbors()
            print("Municipality informs neighbors")
            self.model.years_after_threshold_breach += 1
        elif self.model.years_after_threshold_breach == 2:
            self.model.years_after_threshold_breach += 1
            print("Municipality still legislating")

        else: # after 3 years legislating, the municipality punishes
            self.enforce()
            print("Municipality enforces")

    def all_households(self):
        """TODO this does not need to be a function"""
        return filter(lambda x: not isinstance(x, Municipality),
                                self.model.schedule.agent_buffer(shuffled=True)
                                )

    def inform_neighbors(self):
        for agent in self.all_households():
            if agent.pollutes and agent.willingness_to_upgrade >= 0.90:
                agent.willingness_to_upgrade = 1

    def enforce(self):
        for agent in self.all_households():
            if agent.pollutes:
                new_willingness = agent.willingness_to_upgrade * self.model.enforcement_influence
                if new_willingness < 1:
                    agent.willingness_to_upgrade = new_willingness
                else:
                    agent.willingness_to_upgrade = 1

class Household(Agent):
    def __init__(self, unique_id, pollutes, willingness, model):
        super().__init__(unique_id, model)
        self.pollutes = pollutes
        self.willingness_to_upgrade = willingness

    def step(self):
        # the random scheduler makes this randomized
        if self.pollutes:
            self.pollute()

            if self.willingness_to_upgrade == 1:
                self.upgrade_oss()
                self.inform_neighbors()


    def inform_neighbors(self):
        neighbors = self.model.grid.get_neighbors(self.pos,
                                                  moore=True,
                                                  include_center=False,
                                                  radius=self.model.neighbour_distance)
        household_neighbors = filter(
            lambda x: not isinstance(x, Municipality), neighbors)

        for agent in household_neighbors:
            if agent.pollutes:
                new_willingness = agent.willingness_to_upgrade * self.model.social_influence
                if new_willingness < 1:
                    agent.willingness_to_upgrade = new_willingness
                else:
                    agent.willingness_to_upgrade = 1

    def pollute(self):
        self.model.total_pollution += self.model.household_pollution

    def upgrade_oss(self):
        self.pollutes = False


class SocialModel(Model):
    def __init__(self,
                 n_agents=100,
                 x_grid=20,
                 y_grid=20,
                 neighbour_distance = 2,
                 pike_threshold = 400,
                 initial_pike_abundance = 390.,
                 household_pollution = 30,
                 social_influence = 1.2,
                 enforcement_influence = 1.4):
        super().__init__()
        self.schedule = RandomActivation(self)
        # if torus=True, top and bottom borders (and left-righ) are connected
        self.grid = SingleGrid(x_grid, y_grid, torus=False)
        self.neighbour_distance = neighbour_distance
        self.pike_threshold = pike_threshold
        self.pike_abundance = initial_pike_abundance
        self.household_pollution = household_pollution
        self.social_influence = social_influence
        self.enforcement_influence = enforcement_influence
        self.total_pollution = 100
        self.years_after_threshold_breach = 0

        # data collection
        self.dc = DataCollector(
            model_reporters={"pollution": lambda m: m.total_pollution,
                             "pike_abundance": lambda m: m.pike_abundance})

        for i in range(n_agents):
            a = Household(i, True, self.random.random(), self)
            self.schedule.add(a)
            self.grid.position_agent(a)

        m = Municipality(n_agents + 1, False, self)
        self.schedule.add(m)
        self.grid.position_agent(m)

    def step(self):
        self.total_pollution = 0
        self.schedule.step()
        self.dc.collect(self)



model = SocialModel()
for i in range(50):
    model.step()

model_df = model.dc.get_model_vars_dataframe()

print(model_df)

File: social_model.jl
using Agents
using Random

@agent Household GridAgent{2} begin
    pollutes::Bool
    willingness_to_upgrade::Float64
end

mutable struct Municipality <: AbstractAgent
    id::Int
    pos::NTuple{2, Int}
    has_legislated::Bool
end

Base.@kwdef mutable struct Parameters
	num_households::Int32 = 100
	neighbour_distance::Float32 = 2.0
	griddims ::Tuple{Vararg{Int64, 2}} = (20, 20)
    pike_threshold::Float32 = 400
    years_after_threshold_breach::Int32 = 0
    total_pollution::Int32 = 0
    pike_abundance::Float32 = 390.
    household_pollution::Float32 = 30
    social_influence::Float32 = 1.2
    enforcement_influence::Float32 = 1.4
    seed::Int32 = 1234
end

properties = Parameters()

function initialize_model(; properties = properties )

    space = GridSpaceSingle(properties.griddims, periodic = false)

    rng = Random.MersenneTwister(properties.seed)

    agents_types = Union{Household, Municipality}

    model = ABM(agents_types, space;
                properties = properties,
                rng = rng,
                scheduler = Schedulers.ByType(false, true, agents_types))

    for n in 1:properties.num_households
        household = Household(n, (1,1), true, rand())
        add_agent_single!(household, model)
    end

    municipality = Municipality(properties.num_households + 1, (1,1), false)
    add_agent_single!(municipality, model)

    return model
end

function pollute!(model::ABM, amount)
    model.total_pollution += amount
end

function upgrade_oss!(house::Household)
    house.pollutes = false
end


function update_willingness!(agent::Household, value = 1.5)
    new_willingness = agent.willingness_to_upgrade * value
    agent.willingness_to_upgrade = new_willingness >= 1 ? 1 : new_willingness
end

function inform_neighbours(agent::Household, model)
    all_neighbours = nearby_agents(agent, model, model.neighbour_distance)
    polluting_neighbours = filter(x -> isa(x, Household) && x.pollutes, collect(all_neighbours))

    if !isempty(polluting_neighbours)
        #neighbour = rand(polluting_neighbours)
        for neighbour in polluting_neighbours
            update_willingness!(neighbour, model.social_influence)
        end
    end
end

function inform_neighbours(::Municipality, model)
    """municipality informs, but only those with an already high willingnes to
    upgrade will actually upgrade their systems"""
    num_households = nagents(model) - 1
    # inform all agents but Municipality
    for id in shuffle(1:num_households)
        neighbour = model[id]
        if neighbour.pollutes && neighbour.willingness_to_upgrade >= 0.90
            neighbour.willingness_to_upgrade = 1
        end
    end
end

"""inform all neighbours randomly"""
function enforce(::Municipality, model)
    """they foresee sanctions, so households increase their willingness to
    upgrade"""
    num_households = nagents(model) - 1
    # inform all agents but Municipality
    for id in shuffle(1:num_households)
        household = model[id]
        if household.pollutes
            update_willingness!(household, model.social_influence)
        end
    end
end


function agent_step!(agent::Household, model::ABM)
    if agent.pollutes
        pollute!(model, model.household_pollution)

        if agent.willingness_to_upgrade == 1
            upgrade_oss!(agent)
            # social engacement scenario
            inform_neighbours(agent, model)
        end
    end
end


function agent_step!(municipality::Municipality, model::ABM)
    if model.pike_abundance <= model.pike_threshold
        if model.years_after_threshold_breach == 0 && !municipality.has_legislated
            println("Municipality starts legislation process")
            model.years_after_threshold_breach += 1
            municipality.has_legislated = true
        end
    end

    if model.years_after_threshold_breach == 1
        inform_neighbours(municipality, model)
        model.years_after_threshold_breach += 1
        println("Municipality informs neighbors")
    elseif model.years_after_threshold_breach == 2
        model.years_after_threshold_breach += 1
        println("Municipality still legislating")
    elseif model.years_after_threshold_breach == 3
        # after 3 years legislating, the municipality punishes
        enforce(municipality, model)
        println("Municipality enforces")
    end
end

function model_step!(model::ABM)
    model.total_pollution = 0
end
